### DISCUSSION TITLE
Interrupts - NMI and Priorities on the RP2040

### DISCUSSION THREAD

**Discussion by robzed** [tabemann/zeptoforth#57] [Q&A] [answered]
I'm looking at separating some high priority interrupts from normal interrupts.

**Question 1** 

0 is the highest priority, and 3 is the lowest.

Most interrupt are 0, apart from these two:
  PendSV 3 
  systick 1 

Since the task scheduler runs off systick, it appears that would be more logical making 2 or 3 - giving at least two clear interrupt priorities for the users - by default. In fact, right now I can't imagine a situation where interrupts are not more important than the task scheduling - so normally I see it a the lowest priority (3). 

Perhaps you have a different opinion or reason?

**Question 2**

Even if it's deemed appropriate for Zeptoforth having systick at priority 1, would anything bad if I moved it to 2 or 3? I don't think I have enough Interrupt priorities to make the system work as I want.

(I think I probably need to move the UART interrupt down from 0 as well, but I'm happy with the consequences of this)

**Question 3 and 4**

Is PendSV used by Zeptoforth? ("In an OS environment, use PendSV for context switching when no other exception is active.")

Is SVCall (or svc) used by Zeptoforth? ("In an OS environment, applications can use SVC instructions to access OS kernel functions and device drivers.")

 
**Question 5**

Also regarding NMI from the RP2040 docs: "Dedicated Non-Maskable Interrupt (NMI) input (which can be driven from any standard interrupt source)" ... this obviously provides another effective priority level above 0.

Is there anything special about writing a handler in Zeptoforth over a normal maskable interrupt ISR?

**Comment by tabemann** [comment, marked as answer]
PendSV is where the multitasker and interrupt bottom halves live, which has a priority of 3 for a reason - it is supposed to be the lowest priority interrupt possible. Note that the SysTick handler increments the SysTick counter and, when appropriate, schedules the PendSV, but it itself does not encompass the multitasker or the interrupt bottom halves; under many circumstances it will fire significantly more frequently than the PendSV handler, but at the same time, the PendSV can be triggered independent of the SysTick (by executing `pause`).

As for moving the SysTick to priority 2, that probably wouldn't break anything, but is not ideal because it would make it impossible to make interrupts that have lower priority than SysTick but higher priority than PendSV. Note that PendSV ought to have the lowest priority of all interrupts, hence why it and only it has priority 3.

As for SVCall, it is not currently used by zeptoforth, but it was used in the past for rather misguided reasons (I wanted to capture the interrupt context for configuring new tasks, before I had figured out how to create my own interrupt context out of whole cloth, which turned out to be far simpler anways).

As for NMI, there is nothing stopping the user from registering an interrupt handler for NMI ─ I just have never had a need for NMI.

**Reply by robzed** [reply]
A big thank you for your detailed answer. It appears that the Zeptoforth scheduler is less basic than I imagined. I will have to study it in more depth!

> SysTick to priority 2 [...]  because it would make it impossible to make interrupts that have lower priority than SysTick but higher priority than PendSV. Note that PendSV ought to have the lowest priority of all interrupts, hence why it and only it has priority 3.

I now also understand why PendSV is priority 3. 

Do you have any examples where you'd want an interrupt with lower priority than SysTick but higher than PendSV?
